runner/websocket.go

```diff
@@ -0,0 +1,131 @@
+	proxyConfig := httpproxy.Config{
+		HTTPProxy:  server.Config.HTTPProxy,
+		HTTPSProxy: server.Config.HTTPSProxy,
+		NoProxy:    server.Config.NoProxy,
+	}
+	dialer := websocket.Dialer{
+		ReadBufferSize:  10240,
+		WriteBufferSize: 10240,
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return proxyConfig.ProxyFunc()(req.URL)
+		},
+	}
+	url, _ := url.Parse(server.Config.APIBaseURL + "/v2/snapshots/websocket")
+	if url.Scheme == "http" {
+		url.Scheme = "ws"
+	} else {
+		url.Scheme = "wss"
+	}
+	headers := make(map[string][]string)
+	headers["Pganalyze-Api-Key"] = []string{server.Config.APIKey}
+	headers["Pganalyze-System-Id"] = []string{server.Config.SystemID}
+	headers["Pganalyze-System-Type"] = []string{server.Config.SystemType}
+	headers["Pganalyze-System-Scope"] = []string{server.Config.SystemScope}
+	headers["Pganalyze-System-Id-Fallback"] = []string{server.Config.SystemIDFallback}
+	headers["Pganalyze-System-Type-Fallback"] = []string{server.Config.SystemTypeFallback}
+	headers["Pganalyze-System-Scope-Fallback"] = []string{server.Config.SystemScopeFallback}
+	headers["User-Agent"] = []string{util.CollectorNameAndVersion}
+	conn, response, err := dialer.DialContext(connCtx, url.String(), headers)
+	if err != nil {
+		cancelConn()
+		logger.PrintWarning("Error starting websocket: %s %v", err, response)
+		return
+	}
+	server.WebSocket.Store(conn)
+	server.Pause.Store(&pganalyze_collector.ServerMessage_Pause{Pause: false})
+	// Writer goroutine
+	go func() {
+		for {
+			select {
+			case <-connCtx.Done():
+				socket := server.WebSocket.Swap(nil)
+				if socket != nil {
+					err = socket.Close()
+					if err != nil {
+						logger.PrintWarning("Error closing websocket: %s", err)
+					}
+				}
+				return
+			case snapshot := <-server.SnapshotStream:
+				logger.PrintVerbose("Uploading snapshot to websocket")
+				err = server.WebSocket.Load().WriteMessage(websocket.BinaryMessage, snapshot)
+				if err != nil {
+					logger.PrintError("Error uploading snapshot: %s", err)
+					cancelConn()
+					return
+				}
+			}
+		}
+	}()
+	// Reader goroutine
+	go func() {
+		for {
+			_, compressedData, err := conn.ReadMessage()
+			if err != nil {
+				if !websocket.IsCloseError(err, 1005) { // Normal close event
+					logger.PrintWarning("Error reading from websocket: %s", err)
+				}
+				server.SelfTest.MarkCollectionAspectError(state.CollectionAspectWebSocket, "error starting WebSocket: %s", err)
+				cancelConn()
+				return
+			}
+			server.SelfTest.MarkCollectionAspectOk(state.CollectionAspectWebSocket)
+			var data bytes.Buffer
+			r, err := zlib.NewReader(bytes.NewReader(compressedData))
+			if err != nil {
+				logger.PrintWarning("Error decompressing ServerMessage: %s", err)
+				return
+			}
+			defer r.Close()
+			io.Copy(&data, r)
+			message := &pganalyze_collector.ServerMessage{}
+			err = proto.Unmarshal(data.Bytes(), message)
+			if err != nil {
+				logger.PrintWarning("Error parsing ServerMessage: %s", err)
```

Comment:
New messages would be added to the [protobuf enum](https://github.com/pganalyze/collector/pull/583/files#diff-4cba84f43462652699ada1f4d9e63a7d83cef4bc51554df1f8ee29458e6e6ac0R6-R14), so the message would still be parsed, though presumably would be empty because its contents didn't match a known `oneof` type.